### PR TITLE
[WIP] feat: fix #539 hal support

### LINKED ISSUE #539: Add HAL support

### PR DESCRIPTION
**Pull request by Simperfit**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | will pass
| Fixed tickets | #539 
| License       | MIT
| Doc PR        | todo

This is totally working  with hal : 

Things todo : 
- [ ] Add tests (unit / behat) using halidator
- [ ] refactor if needed ?
- [ ] Drink some coffee 
- [ ] Create doc PR

### CODE REVIEW

File: src/Serializer/ItemNormalizer.php [hunk 1]
```diff
@@ -138,7 +143,7 @@ public function denormalize($data, $class, $format = null, array $context = [])
      */
     protected function extractAttributes($object, $format = null, array $context = [])
     {
-        return [];
+        return ['toto'];
```
Comment (Simperfit) [inline src/Serializer/ItemNormalizer.php]: remove

**Comment by dunglas** [comment]
For error serialization with HAL, we should support one of those formats: http://phlyrestfully.readthedocs.io/en/latest/problems.html

File: src/Bridge/Symfony/Bundle/DependencyInjection/ApiPlatformExtension.php [hunk 2]
```diff
@@ -117,7 +120,7 @@ public function load(array $configs, ContainerBuilder $container)
     }
 
     /**
-     * Enables JSON-LD and Hydra support.
+     * Enables JSON-LD, HAL and Hydra support.
```
Comment (teohhanhui) [inline src/Bridge/Symfony/Bundle/DependencyInjection/ApiPlatformExtension.php]: Why the change here? `hal.xml` is not being loaded here...

File: src/Action/EntrypointAction.php [hunk 3]
```diff
@@ -9,12 +9,12 @@
  * file that was distributed with this source code.
  */
 
-namespace ApiPlatform\Core\Hydra\Action;
+namespace ApiPlatform\Core\Action;
 
 use ApiPlatform\Core\JsonLd\EntrypointBuilderInterface;
```
Comment (teohhanhui) [inline src/Action/EntrypointAction.php]: Shouldn't the namespace here be changed to `ApiPlatform\Core\Hypermedia` too?

File: src/Action/EntrypointAction.php [hunk 4]
```diff
@@ -9,12 +9,12 @@
  * file that was distributed with this source code.
  */
 
-namespace ApiPlatform\Core\Hydra\Action;
+namespace ApiPlatform\Core\Action;
```
Comment (teohhanhui) [inline src/Action/EntrypointAction.php]: Perhaps `ApiPlatform\Core\Hypermedia\Action`?

File: src/Bridge/Symfony/Validator/Hal/EventListener/ValidationExceptionListener.php [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+{
+    private $normalizer;
+
+    public function __construct(NormalizerInterface $normalizer)
+    {
+        $this->normalizer = $normalizer;
+    }
+
+    /**
+     * Returns a list of violations normalized in the Hydra format.
```
Comment (teohhanhui) [inline src/Bridge/Symfony/Validator/Hal/EventListener/ValidationExceptionListener.php]: not Hydra

File: src/EventListener/ExceptionListener.php [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+ * Handle requests errors.
+ *
+ * @author Samuel ROZE <samuel.roze@gmail.com>
+ * @author Kévin Dunglas <dunglas@gmail.com>
+ */
+final class ExceptionListener extends BaseExceptionListener
+{
+    public function onKernelException(GetResponseForExceptionEvent $event)
+    {
+        // Normalize exceptions with hydra errors only for resources
```
Comment (teohhanhui) [inline src/EventListener/ExceptionListener.php]: not hydra

File: src/Hal/Serializer/CollectionFiltersNormalizer.php [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+
+    /**
+     * {@inheritdoc}
+     */
+    public function normalize($object, $format = null, array $context = [])
+    {
+        $data = $this->collectionNormalizer->normalize($object, $format, $context);
+        $resourceClass = $this->resourceClassResolver->getResourceClass($object, $context['resource_class'] ?? null, true);
+
+        if (isset($context['jsonld_sub_level'])) {
```
Comment (teohhanhui) [inline src/Hal/Serializer/CollectionFiltersNormalizer.php]: Maybe the Serializer context keys should be renamed?

* `api_has_context`
* `api_resource_class`
* `api_sub_level`

File: src/Action/EntrypointAction.php (hunk 4 above)
Comment (dunglas) [inline src/Action/EntrypointAction.php]: IMO we can get rid of the `Hypermedia` namespace. The entrypoint can be useful even in a non-hypermedia context (i.e raw JSON or XML).

Comment (teohhanhui) [inline src/Action/EntrypointAction.php]: Not sure what an "entrypoint" is supposed to be (outside of the Hydra context), actually. It's a Hydra thing. Perhaps we shouldn't try to shoehorn things...

Comment (dunglas) [inline src/Action/EntrypointAction.php]: HAL as a similar concept. And it's a common use case to list all resources collection exposed by the API.

Comment (dunglas) [inline src/Action/EntrypointAction.php]: Schema.org itself contains this concept: https://schema.org/EntryPoint

Comment (teohhanhui) [inline src/Action/EntrypointAction.php]: > Schema.org itself contains this concept: https://schema.org/EntryPoint

No. That's a different thing, if you look carefully. Their naming is just confusing sometimes... :laughing:

Comment (dunglas) [inline src/Action/EntrypointAction.php]: Good catch. But HAL has it: http://haltalk.herokuapp.com/explorer/browser.html#/

Comment (Simperfit) [inline src/Action/EntrypointAction.php]: I understand as is, an entrypoint is the /, the base of the app. The first thing you will see (even if there is authentification, you will be redirect or not allowed).

File: src/EventListener/ExceptionListener.php (hunk 6 above)
Comment (dunglas) [inline src/EventListener/ExceptionListener.php]: Not HAL as well, should be removed from this PR.
